espflash/src/partition_table.rs

```diff
@@ -56,58 +56,113 @@ impl Type {
     }
 }
 
-impl Display for Type {
+impl Display for CoreType {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
         write!(f, "{}", serde_plain::to_string(self).unwrap())
     }
 }
 
-#[derive(Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq, BinRead)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, BinRead, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum Type {
+    CoreType(CoreType),
+    Custom(u8),
+}
+
+impl Type {
+    pub fn subtype_hint(&self) -> String {
+        match self {
+            Type::CoreType(ty) => ty.subtype_hint(),
+            Type::Custom(_) => "0x00-0xFE".into(),
+        }
+    }
+
+    pub fn as_u8(&self) -> u8 {
+        match self {
+            Type::CoreType(ty) => *ty as u8,
+            Type::Custom(ty) => *ty,
+        }
+    }
+}
+
+impl Display for Type {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        let s = match *self {
+            Type::CoreType(ty) => serde_plain::to_string(&ty).unwrap(),
+            Type::Custom(ty) => format!("{:#04x}", ty),
+        };
+
+        write!(f, "{}", s)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq, BinRead, EnumString)]
 #[repr(u8)]
 #[br(little, repr = u8)]
 pub enum AppType {
     #[serde(rename = "factory")]
+    #[strum(serialize = "factory")]
     Factory = 0x00,
     #[serde(rename = "ota_0")]
+    #[strum(serialize = "ota_0")]
     Ota0 = 0x10,
```

Comment:
I'm currently using `strum_macros::EnumString` to convert from a `String` to an `AppType` discriminant. I unfortunately cannot use `#[strum(serialize_all = "snake_case")]` here as we need to insert the underscores into the serialized names.

Is there any way I can perform this conversion using `serde` instead? I'm not very experienced with it.

Reply:
Something like this?

 ```rust
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Copy, Clone, BinRead)]
#[serde(untagged)]
pub enum SubType {
    App(AppType),
    Data(DataType),
    #[serde(deserialize_with = "deserialize_custom_partition_sub_type")]
    Custom(u8),
}
```
`untagged` already handles that "first successful parse" logic for you.

Reply:
You're right, I was able to re-work and simplify this using this approach. Thanks!